dataclients/kubernetes/endpointslices.go

```diff
@@ -0,0 +1,154 @@
+package kubernetes
+
+import (
+	"github.com/zalando/skipper/dataclients/kubernetes/definitions"
+)
+
+// There are [1..N] Kubernetes endpointslices created for a single Kubernetes service.
+// Kubernetes endpointslices of a given service can have duplicates with different states.
+// Therefore Kubernetes endpointslices need to be de-duplicated before usage.
+// The business object skipperEndpointSlice is a de-duplicated endpoint list that concats all endpointslices of a given service into one slice of skipperEndpointSlice.
+type skipperEndpointSlice struct {
+	Meta      *definitions.Metadata
+	Endpoints []*skipperEndpoint
+	Ports     []*endpointSlicePort
+}
+
+// Conditions have to be evaluated before creation
+type skipperEndpoint struct {
+	Address string
+	Zone    string
+}
+
+func (eps *skipperEndpointSlice) getPort(protocol, pName string, pValue int) int {
+	var port int
+	for _, p := range eps.Ports {
+		if protocol != "" && p.Protocol != protocol {
+			continue
+		}
+		if pName != "" && p.Name == pName {
+			port = p.Port
+			break
+		}
+		if pValue != 0 && p.Port == pValue {
+			port = pValue
+			break
+		}
+	}
+
+	return port
+}
+func (eps *skipperEndpointSlice) targetsByServicePort(protocol, backendProtocol string, servicePort *servicePort) []string {
+	var port int
+	if servicePort.Name != "" {
+		port = eps.getPort(protocol, servicePort.Name, servicePort.Port)
+	} else if servicePort.TargetPort != nil {
+		var ok bool
+		port, ok = servicePort.TargetPort.Number()
+		if !ok {
+			port = eps.getPort(protocol, servicePort.Name, servicePort.Port)
+		}
+	} else {
+		port = eps.getPort(protocol, servicePort.Name, servicePort.Port)
+	}
+
+	var result []string
+	for _, ep := range eps.Endpoints {
+		result = append(result, formatEndpointString(ep.Address, backendProtocol, port))
+	}
+
+	return result
+}
+
+func (eps *skipperEndpointSlice) targetsByServiceTarget(protocol, backendProtocol string, serviceTarget *definitions.BackendPort) []string {
+	pName, _ := serviceTarget.Value.(string)
+	pValue, _ := serviceTarget.Value.(int)
+	port := eps.getPort(protocol, pName, pValue)
+
+	var result []string
+	for _, ep := range eps.Endpoints {
+		result = append(result, formatEndpointString(ep.Address, backendProtocol, port))
+	}
+
+	return result
+}
+
+func (eps *skipperEndpointSlice) targets(protocol, backendProtocol string) []string {
+	result := make([]string, 0)
```

Comment:
We can pre-allocate capacity of len(eps.Endpoints) here

Reply:
done